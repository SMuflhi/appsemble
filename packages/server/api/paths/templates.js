export default {
  '/api/templates': {
    get: {
      tags: ['template'],
      description: 'Fetch a list of all available templates.',
      operationId: 'getAppTemplates',
      responses: {
        200: {
          description: 'The list of all available templates.',
          content: {
            'application/json': {
              schema: {
                type: 'array',
                items: {
                  schema: {
                    type: 'object',
                    properties: {
                      template: {
                        type: 'string',
                        description: 'The name of the template.',
                      },
                      description: {
                        type: 'string',
                        description: 'The description of the template.',
                      },
                      resources: {
                        type: 'boolean',
                        description: 'Whether this template supports pre-made resources',
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
    post: {
      tags: ['template'],
      description: 'Register a new app using a template.',
      operationId: 'createTemplateApp',
      requestBody: {
        description: 'The template to use for app creation.',
        content: {
          'application/json': {
            schema: {
              type: 'object',
              required: ['template', 'organizationId'],
              properties: {
                template: {
                  type: 'string',
                  description: 'The name of the template.',
                },
                name: {
                  $ref: '#/components/schemas/App/properties/name',
                },
                description: {
                  $ref: '#/components/schemas/App/properties/description',
                },
                organizationId: {
                  $ref: '#/components/schemas/Organization/properties/id',
                },
              },
            },
          },
        },
      },
      responses: {
        201: {
          $ref: '#/components/responses/app',
        },
      },
      security: [{ apiUser: ['apps:write'] }],
    },
  },
};
