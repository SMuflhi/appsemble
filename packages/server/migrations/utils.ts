import { randomUUID } from 'node:crypto';

import { logger } from '@appsemble/node-utils';
import { Op, QueryTypes, type Sequelize } from 'sequelize';

import { App, AppMember, EmailAuthorization, User } from '../models/index.js';

export async function convertUserToAppMember(
  db: Sequelize,
  appId: number,
  userId: string,
): Promise<string> {
  if (!userId) {
    return;
  }

  const app = appId ? await App.findByPk(appId, { attributes: ['id', 'definition'] }) : null;

  if (!app) {
    return;
  }

  const user = await User.findByPk(userId, {
    attributes: ['name', 'primaryEmail', 'locale'],
    include: [{ model: EmailAuthorization }],
  });

  const appMember = await AppMember.findOne({
    where: { AppId: appId, [Op.or]: [{ UserId: userId }, { email: user.primaryEmail }] },
    attributes: ['id'],
  });

  let memberId = appMember ? appMember.id : null;

  // App needs to have a security definition to assign a role to an AppMember
  if (!app.definition.security) {
    app.definition.security = {
      default: {
        role: 'User',
        policy: 'everyone',
      },
      roles: {
        User: { description: 'Default User role created through migration.' },
      },
    };
  }

  if (!memberId) {
    logger.warn(`AppMember account not found for user ${userId}, creating new app account`);
    memberId = randomUUID();
    await db.query(
      `
        INSERT INTO "AppMember" (id, "UserId", "AppId", name, role, email, "emailVerified", locale, created, updated) \
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, NOW(), NOW())
      `,
      {
        type: QueryTypes.INSERT,
        replacements: [
          memberId,
          userId,
          app.id,
          user?.name,
          app.definition.security?.default.role || 'User',
          user?.primaryEmail,
          user?.EmailAuthorizations?.[0]?.verified ?? false,
          user?.locale,
        ],
      },
    );
  }
  return memberId;
}
