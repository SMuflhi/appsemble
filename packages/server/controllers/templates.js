import { normalize } from '@appsemble/utils';
import Boom from '@hapi/boom';
import crypto from 'crypto';
import { UniqueConstraintError } from 'sequelize';

import templates from '../templates/apps';
import generateVapidToken from '../utils/generateVapidToken';
import getAppFromRecord from '../utils/getAppFromRecord';

export async function getAppTemplates(ctx) {
  ctx.body = templates.map(({ name, description, resources }) => ({
    name,
    description,
    resources: !!resources,
  }));
}

export async function createTemplateApp(ctx) {
  const {
    template: reqTemplate,
    name,
    description,
    organizationId,
    resources,
    private: isPrivate = true,
  } = ctx.request.body;
  const { App, Resource } = ctx.db.models;
  const { user } = ctx.state;

  const template = templates.find(t => t.name === reqTemplate);

  if (!user.organizations.some(organization => organization.id === organizationId)) {
    throw Boom.forbidden('User does not belong in this organization.');
  }

  if (!template) {
    throw Boom.notFound(`Template ${template} does not exist.`);
  }

  try {
    const path = name ? normalize(name) : normalize(template);
    const result = {
      definition: {
        ...template.definition,
        description,
        name: name || template,
        private: isPrivate,
      },
      OrganizationId: organizationId,
      path,
      ...(resources && {
        Resources: [].concat(
          ...Object.keys(template.resources).map(key =>
            template.resources[key].map(r => ({ type: key, data: r })),
          ),
        ),
      }),
    };

    let record;
    for (let i = 2; i < 12; i += 1) {
      try {
        // eslint-disable-next-line no-await-in-loop
        record = await App.create(result, { include: [Resource] });

        if (record) {
          break;
        }
      } catch (ex) {
        if (ex instanceof UniqueConstraintError) {
          result.path = `${path}-${i}`;
        } else {
          throw ex;
        }
      }
    }

    if (!record) {
      // Fallback if a suitable ID could not be found after trying for a while
      result.path = `${path}-${crypto.randomBytes(5).toString('hex')}`;
      record = await App.create(result, { include: [Resource] });
    }

    await record.createAppNotificationKey(generateVapidToken());

    ctx.body = getAppFromRecord(record);
    ctx.status = 201;
  } catch (error) {
    if (error instanceof UniqueConstraintError) {
      throw Boom.conflict(
        `Another app with path “${name ? normalize(name) : normalize(template)}” already exists`,
      );
    }

    throw error;
  }
}
