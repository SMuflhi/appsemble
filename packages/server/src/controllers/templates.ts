import { normalize, Permission } from '@appsemble/utils';
import Boom from '@hapi/boom';
import crypto from 'crypto';
import { col, fn, UniqueConstraintError } from 'sequelize';
import { generateVAPIDKeys } from 'web-push';

import { App, Resource } from '../models';
import type { KoaContext } from '../types';
import checkRole from '../utils/checkRole';
import getAppFromRecord from '../utils/getAppFromRecord';

export async function getAppTemplates(ctx: KoaContext): Promise<void> {
  const templates = await App.findAll({
    where: { template: true },
    attributes: {
      include: ['id', 'definition', [fn('COUNT', col('Resources.id')), 'ResourceCount']],
    },
    include: [{ model: Resource, attributes: [] }],
    group: ['App.id'],
  });

  ctx.body = templates.map((template) => ({
    id: template.id,
    name: template.definition.name,
    description: template.definition.description,
    resources: template.get('ResourceCount') > 0,
  }));
}

export async function createTemplateApp(ctx: KoaContext): Promise<void> {
  const {
    request: {
      body: { description, name, organizationId, private: isPrivate, resources, templateId },
    },
  } = ctx;

  const template = await App.findOne({
    where: { id: templateId },
    include: [Resource],
  });

  await checkRole(ctx, organizationId, Permission.CreateApps);

  if (!template) {
    throw Boom.notFound(`Template with ID ${templateId} does not exist.`);
  }

  if (!template.template && template.private) {
    await checkRole(ctx, template.OrganizationId, Permission.ViewApps);
  }

  const path = name ? normalize(name) : normalize(template.definition.name);
  try {
    const keys = generateVAPIDKeys();
    const result: Partial<App> = {
      definition: {
        ...template.definition,
        description,
        name: name || template,
      },
      private: Boolean(isPrivate),
      vapidPublicKey: keys.publicKey,
      vapidPrivateKey: keys.privateKey,
      OrganizationId: organizationId,
      ...(resources && {
        Resources: [].concat(template.Resources.map(({ data, type }) => ({ type, data }))),
      }),
    };

    for (let i = 1; i < 11; i += 1) {
      const p = i === 1 ? path : `${path}-${i}`;
      const count = await App.count({ where: { path: p } });
      if (count === 0) {
        result.path = p;
        break;
      }
    }

    if (!result.path) {
      // Fallback if a suitable ID could not be found after trying for a while
      result.path = `${path}-${crypto.randomBytes(5).toString('hex')}`;
    }

    const record = await App.create(result, { include: [Resource] });

    ctx.body = getAppFromRecord(record);
    ctx.status = 201;
  } catch (error) {
    if (error instanceof UniqueConstraintError) {
      throw Boom.conflict(`Another app with path “${path}” already exists`);
    }

    throw error;
  }
}
