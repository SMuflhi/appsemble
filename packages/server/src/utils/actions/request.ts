import { RequestLikeActionDefinition } from '@appsemble/types';
import { defaultLocale, formatRequestAction, remap } from '@appsemble/utils';
import axios from 'axios';

import { ServerActionParameters } from '.';
import { getRemapperContext } from '../app';

export async function request({ action, app, data, user }: ServerActionParameters): Promise<any> {
  let method: 'DELETE' | 'GET' | 'POST' | 'PUT';

  if (!(action as RequestLikeActionDefinition).method) {
    switch (action.type) {
      case 'resource.update':
        method = 'PUT';
        break;
      case 'resource.delete':
        method = 'DELETE';
        break;
      case 'resource.create':
        method = 'POST';
        break;
      default:
        method = 'GET';
    }
  }

  const context = await getRemapperContext(
    app,
    app.definition.defaultLanguage || defaultLocale,
    user && {
      sub: user.id,
      name: user.name,
      email: user.primaryEmail,
      email_verified: Boolean(user.EmailAuthorizations?.[0]?.verified),
    },
  );
  const axiosConfig = formatRequestAction(
    { ...action, method },
    data,
    (remapper, d) => remap(remapper, d, context),
    context.context,
  );
  const response = await axios(axiosConfig);

  return response.data;
}
