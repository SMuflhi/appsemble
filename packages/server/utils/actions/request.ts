import { getRemapperContext } from '@appsemble/node-utils/app';
import {
  Remapper,
  RequestLikeActionDefinition,
  ResourceQueryActionDefinition,
} from '@appsemble/types';
import { defaultLocale, formatRequestAction, remap } from '@appsemble/utils';
import axios from 'axios';

import { ServerActionParameters } from './index.js';

export async function request({
  action,
  app,
  context,
  data,
  options,
  user,
}: ServerActionParameters): Promise<any> {
  let method: 'DELETE' | 'GET' | 'POST' | 'PUT';
  const definition = action as RequestLikeActionDefinition;
  const query: Remapper = []
    .concat(
      definition?.query ?? action.type.startsWith('resource.')
        ? app.definition.resources[(action as ResourceQueryActionDefinition).resource]?.query?.query
        : undefined,
    )
    .filter(Boolean);

  if (!(action as RequestLikeActionDefinition).method) {
    switch (action.type) {
      case 'resource.update':
        method = 'PUT';
        break;
      case 'resource.delete':
        method = 'DELETE';
        break;
      case 'resource.create':
        method = 'POST';
        break;
      default:
        method = 'GET';
    }

    if ((action.type === 'resource.get' || action.type === 'resource.query') && action.view) {
      query.push({ 'object.assign': { view: action.view } });
    }
  }

  const remapperContext = await getRemapperContext(
    app.toJSON(),
    app.definition.defaultLanguage || defaultLocale,
    user && {
      sub: user.id,
      name: user.name,
      email: user.primaryEmail,
      email_verified: Boolean(user.EmailAuthorizations?.[0]?.verified),
      zoneinfo: user.timezone,
    },
    options,
    context,
  );
  const axiosConfig = formatRequestAction(
    { ...action, query: query.length ? query : undefined, method },
    data,
    (remapper, d) => remap(remapper, d, remapperContext),
    context.context,
  );
  const response = await axios(axiosConfig);

  return response.data;
}
