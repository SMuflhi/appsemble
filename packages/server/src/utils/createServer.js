import { loggerMiddleware } from '@appsemble/node-utils';
import faPkg from '@fortawesome/fontawesome-free/package.json';
import Boom from '@hapi/boom';
import cors from '@koa/cors';
import Koa from 'koa';
import compose from 'koa-compose';
import compress from 'koa-compress';
import mount from 'koa-mount';
import koaQuerystring from 'koa-qs';
import range from 'koa-range';
import session from 'koa-session';
import serve from 'koa-static';
import koasBodyParser from 'koas-body-parser';
import koas from 'koas-core';
import koasOperations from 'koas-operations';
import koasParameters from 'koas-parameters';
import koasSecurity from 'koas-security';
import koasSerializer from 'koas-serializer';
import koasSpecHandler from 'koas-spec-handler';
import koasStatusCode from 'koas-status-code';
import koasSwaggerUI from 'koas-swagger-ui';
import path from 'path';
import raw from 'raw-body';

import api from '../api';
import * as operations from '../controllers';
import appMapper from '../middleware/appMapper';
import boom from '../middleware/boom';
import conditional from '../middleware/conditional';
import frontend from '../middleware/frontend';
import oauth2 from '../middleware/oauth2';
import tinyRouter from '../middleware/tinyRouter';
import { appRouter, studioRouter } from '../routes';
import bulmaHandler from '../routes/bulmaHandler';
import authentication from './authentication';
import convertToCsv from './convertToCsv';
import Mailer from './email/Mailer';

export default async function createServer({ app = new Koa(), argv = {}, db, webpackConfigs }) {
  // eslint-disable-next-line no-param-reassign
  app.keys = [argv.secret];
  app.use(loggerMiddleware());
  app.use(session(app));

  app.use(boom());
  app.use(range);
  Object.assign(app.context, { argv, db, mailer: new Mailer(argv) });

  koaQuerystring(app);

  if (process.env.NODE_ENV === 'production') {
    app.use(compress());
  }

  app.use(
    mount(
      `/fa/${faPkg.version}`,
      serve(path.dirname(require.resolve('@fortawesome/fontawesome-free/package.json'))),
    ),
  );

  app.use(
    tinyRouter([
      {
        route: /^\/bulma/,
        get: bulmaHandler,
      },
    ]),
  );

  const apiMiddleware = mount(
    '/api',
    compose([
      await koas(api(), [
        koasSpecHandler(),
        koasSwaggerUI({ url: '/explorer' }),
        koasSecurity(authentication(argv, db.models)),
        () => (ctx, next) => {
          if (ctx.users) {
            [ctx.state.user] = Object.values(ctx.users);
          }
          return next();
        },
        koasParameters(),
        koasBodyParser({
          '*/*': (body, mediaTypeObject, ctx) =>
            raw(body, {
              length: ctx.request.length,
            }),
        }),
        koasSerializer({
          'text/csv': body => convertToCsv(body),
        }),
        koasStatusCode(),
        koasOperations({ operations }),
      ]),
      () => {
        throw Boom.notFound('URL not found');
      },
    ]),
  );

  if (process.env.NODE_ENV !== 'test') {
    app.use(await frontend(webpackConfigs));
  }

  app.use(
    appMapper(
      compose([
        conditional(ctx => ctx.path.startsWith('/api') || ctx.path === '/oauth2/token', cors()),
        apiMiddleware,
        studioRouter,
        oauth2(argv),
      ]),
      compose([apiMiddleware, appRouter]),
    ),
  );

  return app.callback();
}
