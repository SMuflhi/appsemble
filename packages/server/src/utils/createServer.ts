import { dirname } from 'path';
import { parse } from 'querystring';

import { loggerMiddleware } from '@appsemble/node-utils';
import { api } from '@appsemble/utils';
import faPkg from '@fortawesome/fontawesome-free/package.json';
import { notFound } from '@hapi/boom';
import cors from '@koa/cors';
import { Readable } from 'form-data';
import Koa from 'koa';
import compose from 'koa-compose';
import compress from 'koa-compress';
import mount from 'koa-mount';
import range from 'koa-range';
import serve from 'koa-static';
import { bodyParser, bufferParser, formdataParser } from 'koas-body-parser';
import { koas } from 'koas-core';
import { operations } from 'koas-operations';
import { parameters } from 'koas-parameters';
import { security } from 'koas-security';
import { serializer } from 'koas-serializer';
import { specHandler } from 'koas-spec-handler';
import { statusCode } from 'koas-status-code';
import { swaggerUI } from 'koas-swagger-ui';
import { OpenAPIV3 } from 'openapi-types';
import { Configuration } from 'webpack';

import * as controllers from '../controllers';
import { appMapper } from '../middleware/appMapper';
import { boomMiddleware } from '../middleware/boom';
import { conditional } from '../middleware/conditional';
import { frontend } from '../middleware/frontend';
import { tinyRouter } from '../middleware/tinyRouter';
import { appRouter, studioRouter } from '../routes';
import { bulmaHandler } from '../routes/bulmaHandler';
import { Argv, KoaContext, KoaMiddleware } from '../types';
import { authentication } from './authentication';
import { convertToCsv } from './convertToCsv';
import { Mailer } from './email/Mailer';
import { readPackageJson } from './readPackageJson';

// @ts-expect-error This is needed due to an upstream bug in Koas
bufferParser.skipValidation = true;
// @ts-expect-error This is needed due to an upstream bug in Koas
formdataParser.skipValidation = true;

async function xWwwFormUrlencodedParser(
  body: Readable,
  mediaTypeObject: OpenAPIV3.MediaTypeObject,
  ctx: KoaContext,
): Promise<any> {
  const buffer = await bufferParser(body, mediaTypeObject, ctx);
  const data = parse(String(buffer));
  return data;
}

interface CreateServerOptions {
  /**
   * The CLI arguments processed by yargs.
   */
  argv: Argv;

  /**
   * Additional middleware to inject before any other middleware.
   *
   * This is used for testing purposes.
   */
  middleware?: KoaMiddleware;

  /**
   * Webpack configurations to serve using Webpack dev server middleware.
   */
  webpackConfigs?: Configuration[];
}

export async function createServer({
  argv = {},
  middleware,
  webpackConfigs,
}: CreateServerOptions): Promise<Koa> {
  const app = new Koa();
  app.keys = [argv.secret];
  app.proxy = argv.proxy;
  if (middleware) {
    app.use(middleware);
  }
  app.use(loggerMiddleware());
  app.use(boomMiddleware());
  app.use(range);
  Object.assign(app.context, { argv, mailer: new Mailer(argv) });

  if (process.env.NODE_ENV === 'production') {
    app.use(compress());
  }

  app.use(
    mount(
      `/fa/${faPkg.version}`,
      serve(dirname(require.resolve('@fortawesome/fontawesome-free/package.json'))),
    ),
  );

  app.use(
    tinyRouter([
      {
        route: /^\/bulma/,
        get: bulmaHandler,
      },
    ]),
  );

  if (process.env.NODE_ENV !== 'test') {
    app.use(await frontend(webpackConfigs));
  }

  app.use(
    appMapper(
      compose([
        conditional((ctx) => ctx.path.startsWith('/api') || ctx.path === '/oauth2/token', cors()),
        await koas(api(readPackageJson().version, argv), [
          specHandler(),
          swaggerUI({ url: '/api-explorer' }),
          security(authentication(argv) as any),
          parameters(),
          bodyParser({
            parsers: {
              'application/x-www-form-urlencoded': xWwwFormUrlencodedParser,
              '*/*': bufferParser,
            },
          }),
          serializer({
            serializers: {
              'application/xml': (body: string) => body,
              'text/csv': convertToCsv,
            },
          }),
          statusCode(),
          operations({ controllers }),
        ]),
        ({ path }, next) => {
          if (path.startsWith('/api/')) {
            throw notFound('URL not found');
          }
          return next();
        },
        studioRouter,
      ]),
      appRouter,
    ),
  );

  return app;
}
