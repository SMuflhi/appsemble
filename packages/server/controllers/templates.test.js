import { validate } from '@appsemble/utils';
import RefParser from 'json-schema-ref-parser';
import jwt from 'jsonwebtoken';
import request from 'supertest';

import schema from '../api';
import templates from '../templates/apps';
import createServer from '../utils/createServer';
import testSchema from '../utils/test/testSchema';
import testToken from '../utils/test/testToken';
import truncate from '../utils/test/truncate';

describe('App Templates', () => {
  templates.map(template =>
    it(`should validate ${template.name}`, async () => {
      const spec = await RefParser.dereference(schema());
      const appSchema = spec.components.schemas.App;
      const result = await validate(appSchema, template.definition);
      expect(result).toBeUndefined();
    }),
  );
});

describe('Template API', () => {
  let db;
  let server;
  let token;
  let organizationId;
  let App;
  let Resource;

  beforeAll(async () => {
    db = await testSchema('assets');

    server = await createServer({ db });
    ({ App, Resource } = db.models);
  }, 10e3);

  beforeEach(async () => {
    await truncate(db);
    token = await testToken(request, server, db, 'apps:read apps:write');
    organizationId = jwt.decode(token.substring(7)).user.organizations[0].id;
  });

  afterAll(async () => {
    await db.close();
  });

  it('should return a list of available templates', async () => {
    const { body: result } = await request(server).get('/api/templates');
    const expected = templates.map(({ name, description, resources }) => ({
      name,
      description,
      resources: !!resources,
    }));

    expect(result).toStrictEqual(expected);
  });

  it('should create a new app using a template', async () => {
    const { body: result } = await request(server)
      .post('/api/templates')
      .set('Authorization', token)
      .send({
        template: templates[0].name,
        name: 'Test app',
        description: 'This is a test app',
        organizationId,
      });

    expect(result).toMatchSnapshot();
  });

  it('should create a new app with example resources', async () => {
    const template = templates.find(t => t.name === 'Person App');
    const { body: result } = await request(server)
      .post('/api/templates')
      .set('Authorization', token)
      .send({
        template: template.name,
        name: 'Test app',
        description: 'This is a test app',
        organizationId,
        resources: true,
      });

    const { id } = result;
    const resources = await Resource.findAll(
      { where: { AppId: id, type: 'person' } },
      { raw: true },
    );

    expect(resources.map(r => r.data)).toStrictEqual(
      expect.arrayContaining(template.resources.person),
    );
  });

  it('should append a number when creating a new app using a template with a duplicate name', async () => {
    await request(server)
      .post('/api/templates')
      .set('Authorization', token)
      .send({
        template: templates[0].name,
        name: 'Test app',
        description: 'This is a test app',
        organizationId,
      });

    const {
      status,
      body: { id },
    } = await request(server)
      .post('/api/templates')
      .set('Authorization', token)
      .send({
        template: templates[0].name,
        name: 'Test app',
        description: 'This is a test app',
        organizationId,
      });
    const { body: settings } = await request(server).get(`/api/apps/${id}/settings`);

    expect(status).toStrictEqual(201);
    expect(settings.path).toStrictEqual('test-app-2');
  });

  it('should fall back to append random bytes to the end of the app path after 10 attempts', async () => {
    for (let i = 1; i < 11; i += 1) {
      // eslint-disable-next-line no-await-in-loop
      await App.create(
        {
          path: i === 1 ? 'test-app' : `test-app-${i}`,
          definition: { name: 'Test App', defaultPage: 'Test Page' },
          OrganizationId: organizationId,
        },
        { raw: true },
      );
    }

    const {
      status,
      body: { id },
    } = await request(server)
      .post('/api/templates')
      .set('Authorization', token)
      .send({
        template: templates[0].name,
        name: 'Test app',
        description: 'This is a test app',
        organizationId,
      });

    const { body: settings } = await request(server).get(`/api/apps/${id}/settings`);

    expect(status).toStrictEqual(201);
    const regex = /test-app-(\w){10}/;
    expect(regex.test(settings.path)).toBe(true);
  });
});
