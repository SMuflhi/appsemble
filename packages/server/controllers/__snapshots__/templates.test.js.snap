// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Template API should create a new app using a template 1`] = `
Object {
  "defaultPage": "Example Page A",
  "description": "This is a test app",
  "id": 1,
  "name": "Test app",
  "organizationId": "testorganization",
  "pages": Array [
    Object {
      "blocks": Array [
        Object {
          "actions": Object {
            "onClick": Object {
              "to": "Example Page B",
              "type": "link",
            },
          },
          "type": "action-button",
          "version": "0.8.2",
        },
      ],
      "name": "Example Page A",
    },
    Object {
      "blocks": Array [
        Object {
          "actions": Object {
            "onClick": Object {
              "to": "Example Page A",
              "type": "link",
            },
          },
          "type": "action-button",
          "version": "0.8.2",
        },
      ],
      "name": "Example Page B",
    },
  ],
  "path": "test-app",
  "private": true,
  "theme": Object {
    "themeColor": "#000000",
  },
  "yaml": "name: Test app
defaultPage: Example Page A
theme:
  themeColor: '#000000'
pages:
  - name: Example Page A
    blocks:
      - type: action-button
        version: 0.8.2
        actions:
          onClick:
            type: link
            to: Example Page B
  - name: Example Page B
    blocks:
      - type: action-button
        version: 0.8.2
        actions:
          onClick:
            type: link
            to: Example Page A
description: This is a test app
private: true
",
}
`;
